lib/smf/smf.c

```diff
@@ -0,0 +1,308 @@
+		if (tmp_state[0] == NULL) {
+			break;
+		}
+
+		tmp_state[0] = tmp_state[0]->parent;
+
+		/*
+		 * Do not execute a parent state's exit action that has
+		 * a shared ancestry with the target.
+		 */
+		parent = target_parent;
+		while (parent != NULL) {
+			if (tmp_state[0] == parent) {
+				tmp_state[0] = NULL;
+				break;
+			}
+
+			/* Get target state next parent state if it exists */
+			parent = parent->parent;
+		}
+
+		if (tmp_state[0] && tmp_state[0]->exit) {
+			/* Execute exit action */
+			tmp_state[0]->exit(ctx);
+
+			/* No need to continue if terminate was set */
+			if (internal->terminate) {
+				return;
+			}
+		}
+	}
+
+	/* All done executing the exit actions */
+#endif /* (CONFIG_SMF_ANCESTOR_NUM > 0) */
+
+	internal->exit = false;
+
+	/* update the state variables */
+	ctx->previous = ctx->current;
+	ctx->current = target;
+
+#if (CONFIG_SMF_ANCESTOR_NUM > 0)
+	/*
+	 * Prepare to execute all entry actions of the target's
+	 * parent states
+	 */
+
+	tmp_state[CONFIG_SMF_ANCESTOR_NUM - 1] = target->parent;
+
+	/* Get all parent states of the target */
+	for (int i = CONFIG_SMF_ANCESTOR_NUM - 1; i > 0; i--) {
+		if (tmp_state[i] != NULL) {
+			tmp_state[i - 1] = tmp_state[i]->parent;
+		} else {
+			tmp_state[i - 1] = NULL;
+		}
+	}
+
+	/* Get parent state of previous state */
+	last_parent = ctx->previous->parent;
+
+	/* Execute all parent state entry actions in forward order */
+	for (int i = 0; i < CONFIG_SMF_ANCESTOR_NUM; i++) {
+		/* No parent state */
+		if (tmp_state[i] == NULL) {
+			continue;
+		}
+
+		/*
+		 * We only want to execute the target state's parent state
+		 * entry action if it doesn't share a parent state with the
+		 * previous state.
+		 */
+		parent = last_parent;
+		no_execute = false;
+		while (parent != NULL) {
+			if (tmp_state[i] == parent) {
+				no_execute = true;
+				break;
+			}
+
+			/* Get last state's next parent state if it exists */
+			parent = parent->parent;
+		}
+
+		/* Execute parent state's entry */
+		if (!no_execute && tmp_state[i]->entry) {
+			tmp_state[i]->entry(ctx);
+
+			/* No need to continue if terminate was set */
+			if (internal->terminate) {
+				return;
+			}
+		}
+	}
+#endif /* (CONFIG_SMF_ANCESTOR_NUM > 0) */
+
+	/* Now execute the target entry action */
+	if (ctx->current->entry) {
+		ctx->current->entry(ctx);
+		/*
+		 * If terminate was set, it will be handled in the
+		 * smf_run_state function
+		 */
+	}
+}
+
+void smf_set_terminate(struct smf_ctx *ctx, int32_t val)
+{
+	struct internal_ctx * const internal = (void *) ctx->internal;
+
+	internal->terminate = true;
+	ctx->terminate_val = val;
+}
+
+int32_t smf_run_state(struct smf_ctx *const ctx)
+{
+#if (CONFIG_SMF_ANCESTOR_NUM > 0)
+	const struct smf_state *tmp_state;
+#endif
+	struct internal_ctx * const internal = (void *) ctx->internal;
+
+	/* No need to continue if terminate was set */
+	if (internal->terminate) {
+		return ctx->terminate_val;
+	}
+
+#if (CONFIG_SMF_ANCESTOR_NUM > 0)
```

Comment:
Move this block into a separate function.

Reply:
done